### ISSUE TITLE
Expose cache class and allow to inject it

### ISSUE THREAD

**Issue by mantoni** [eslint/eslint#10781]
**The version of ESLint you are using.**

eslint `v5.4.0`

**The problem you want to solve.**

Hi eslint team 👋

I'm the maintainer of [eslint_d](https://github.com/mantoni/eslint_d.js) which aims at accelerating eslint, especially when used frequently on single files, e.g. as an editor plugin.

Looking at the eslint cache implementation, it strikes me that the `eslint_d` background server could keep an in-memory cache instance, making it obsolete to read and write the cache file. To support this, I would need slightly more control over the cache.

**Your take on the correct solution to problem.**

My suggestion is to expose a cache class and allow to inject instances as an option for the `CLIEngine`. If the cache is injected, it would be the callers responsibility to persist the cache.

If you'd consider changing the API for this, I'm happy to contribute a pull request.

Thanks for making eslint!

**Comment by ilyavolodin** [comment]
That's not a bad idea, I like it. This will have to go through TSC for a vote, since this is an enhancement to the core.

**Comment by not-an-aardvark** [comment]
Thanks for creating an issue. Could you elaborate on what kind of API operations you would need in order to make this work? I think it would make sense to provide enough API surface to be able to do this, but ideally it would be nice to provide the minimal API surface possible to ensure that we have the ability to change the implementation later on. For example, we currently have an [internal set of operations](https://github.com/eslint/eslint/blob/cb946affd21e701fc8309fcb3305e5a1b8d285ed/lib/util/lint-result-cache.js) for dealing with lint result cache, but I'm not sure it would be a good idea to expose that whole class, since some of the functionality probably doesn't need to be exposed.

**Comment by mantoni** [comment]
I've made an first attempt to implement this as a base for further discussion.
Here's the diff: https://github.com/eslint/eslint/compare/master...mantoni:cache-api – let me know if you want me to turn this into a PR.

Usage example:

```js
const eslint = require('eslint');

const engine = new eslint.CLIEngine(/* options */);

const cache = engine.createResultCache();
engine.executeOnFiles([ /* files */ ], cache);
```

It turned out that a change is needed in `file-entry-cache` because `reconcile()` always persists the cache to disk. The call to `reconcile()` is still required because that is the only method where the cache is actually populated.

I've made a PR to support skipping the final `cache.save()` call as an option. Without this change, the new API is still working, but always persists the cache to disk.

As suggested in my previous comment, there is more than one way to expose the cache API and I just wanted to get something out as a discussion base. I'm not attached to this solution.

I did some initial performance checks with `time` comparing `eslint` and `eslint_d` with this change applied (linting eslint_d itself):

| Command | Time (real) |
| --- | --- |
| `eslint .` | 0.85 |
| `eslint . --cache` (cold) | 0.87 |
| `eslint . --cache` (warm) |  0.46 |
| `eslint_d .` (no patch, cold) | 0.82  |
| `eslint_d .` (no patch, warm) | 0.34  |
| `eslint_d .` (with patch, cold) | 0.83 |
| `eslint_d .` (with patch, warm) | 0.10 |
| `eslint_d .` (with patch, warm, one file touched) | 0.13 |

**Comment by not-an-aardvark** [comment]
Thanks for working on this!

I'm hesitant to add something like `engine.createResultCache` or the ability to inject a `file-entry-cache` instance, because it seems like would lock us into using `file-entry-cache`'s API indefinitely. In the future, it's possible that we will want to change the cache file format and/or use a different library with a different API. My impression is that the "injection" approach would be easy to implement in the short term but would cause problems in the long term by exposing term because it would expose too much API surface and prevent us from changing the implementation.

To me, it seems like it would be nice to have the API look vaguely like this:

```js
const engine = new eslint.CLIEngine(
    {
        cache: {
            getUpdatedFiles(files) {
                // returns a list of updated files
            },
            updateCache(changesToCache) {
                // makes changes to the cache
            }
        }
    }
);
```

Then `CLIEngine` would call the provided `getUpdatedFiles` and `updateCache` methods while running in order to get data from the cache.

* With the `--cache` flag, the default behavior would be for `getUpdatedFiles` to read `.eslintcache` and check the mtime, and for `updateCache` to write an `.eslintcache` file to the filesystem, as it does now.
* Without the `--cache` flag, the default behavior would be for `getUpdateFiles` to assume all files are updated, and for `updateCache` to be a no-op. This would be the same as the current behavior when running without a cache. However, it would remove some complexity from `CLIEngine#executeOnText` because it would no longer be necessary to have special logic for the `--cache` flag -- the difference in behavior would be encapsulated in the different `getUpdatedFiles` and `updateCache` methods.
* `eslint_d` would provide a `getUpdatedFiles` method that reads information from memory, and an `updateCache` method that writes to memory.

**Comment by mantoni** [comment]
I like your suggestion. It's propably a good idea to keep the cache format and the caching implementation details private.

One question though: How do we provide the initial cache state to eslint? My understanding is that the cache provides eslint with the previous results, plus meta data on when the files where last checked.

Maybe we can leave caching entirely to eslint internals. How about this?

```js
const engine = new eslint.CLIEngine({
  getUpdatedFiles(files) {
    // returns a list of updated files
  }
});
engine.executeOnFiles([ /* patterns */ ], [ /* previous results to update */ ]);
```

The `results` returned by eslint are the only exchange format needed and we don't have to expose an additional cache format.

**Comment by not-an-aardvark** [comment]
The integration would be responsible for maintaining the initial cache state (e.g. by reading an `.eslintcache` file). ESLint wouldn't need to know the internal cache state -- it would just read information about it by calling `getUpdatedFiles` and `updateCache`.

I'm thinking that ESLint core doesn't need to deal with file modification times at all. The process would look like this:

1. The CLI user (or integration) passes a list of filenames to ESLint
1. ESLint calls `getUpdatedFiles` on the list of filenames.
1. The implementation of `getUpdatedFiles` reads `.eslintcache` from the filesystem and compares its stored file modification times to the modification times of files on disk. It returns a list of files that have been modified since the cache was created.
1. ESLint lints the returned files, obtains a set of linting results in some form, and calls `updateCache` on them.
1. The implement of `updateCache` writes results back to the `.eslintcache` file in the filesystem.

A few notes on these steps:

* This would require the implementation of `getUpdatedFiles` to be aware of the `--cache-location` flag, so that it can read/write results to the correct location. Maybe it could take that as an additional argument or something.
* This would make ESLint indifferent to the cache file format; different integrations could have different cache formats.
* Having an integration pass a list of files to ESLint, only for ESLint to pass the files back to the integration for filtering, seems like a fairly roundabout way of doing things. Is there an advantage to doing it this way? Would it work for `eslint_d` to effectively handle caching on its own by checking file modification times and filtering a list of files before passing the files to ESLint?

**Comment by ilyavolodin** [comment]
I would make a small change to the flow that @not-an-aardvark proposes. Instead of having `getUpdatedFiles` function that takes the list of all files, I would do it one file at a time (something like `isFileCached(filepath, content)`). That should speed things up, since we could call that method just before parsing file into AST and instead of reading file twice, we would only have to do that once.

**Comment by mantoni** [comment]
Regarding @not-an-aardvark's last question: `executeOnFiles(pattern)` receives a pattern which is then resolved to a list of files. Therefore the roundabout way of doing things. Since I'm trying to be 100% eslint compatible, I wouldn't want to expand patterns myself. We could expose that separately though, e.g. `engine.listFilesToProcess(pattern)`. With the normalized file list, I might even be able to implement my own results caching without touching the eslint core cache implementation 🤔

**Comment by mantoni** [comment]
Ok, I experimented with a custom cache implementation in `eslint_d`. It turns out that the only change I would need is to expose `listFilesToProcess` from `glob-utils.js`: https://github.com/eslint/eslint/compare/master...mantoni:list-files-to-process

This allows me to obtain the files for patterns as they're matched by eslint. I can then iterate over the list to find modified files. The filtered list is then passed to eslint and the returned results are stored in my cache. The only tiny copy-paste I had to do is `calculateStatsPerRun`. My implementation relies only on the documented `report.results` format now, so nothing else would need to be changed.

I'm not sure anymore whether exposing a cache API is the right approach.

Would it be okay to expose `listFilesToProcess`? If so, I can add a test case and send a PR for this.
